File: 0-video-ingestion-and-transcoding/2-csharp/TranscodeService.cs
using System.Diagnostics;
using Microsoft.EntityFrameworkCore;

namespace TranscodeService;

// Orchestrates job creation, FFmpeg, and MinIO upload.
public class TranscodeWorker
{
    private readonly IServiceScopeFactory _scopeFactory;
    private readonly MinioStorageService _storage;
    private readonly ILogger<TranscodeWorker> _logger;

    public TranscodeWorker(IServiceScopeFactory scopeFactory, MinioStorageService storage, ILogger<TranscodeWorker> logger)
    {
        _scopeFactory = scopeFactory;
        _storage = storage;
        _logger = logger;
    }

    // CreateJobAsync: insert a PENDING job row and fire FFmpeg in the background.
    // Returns immediately with HTTP 202 — the transcode is NOT awaited here.
    public async Task<TranscodeJob> CreateJobAsync(string inputPath)
    {
        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        // Idempotency: same inputPath (not FAILED) -> return the existing job.
        var existing = await db.TranscodeJobs
            .FirstOrDefaultAsync(j => j.InputPath == inputPath && j.Status != TranscodeStatus.Failed);
        if (existing is not null) return existing;

        var job = new TranscodeJob { InputPath = inputPath };
        db.TranscodeJobs.Add(job);
        await db.SaveChangesAsync();

        // Fire-and-forget: run FFmpeg without blocking the HTTP response.
        _ = Task.Run(() => RunTranscodeAsync(job.Id));
        return job;
    }

    private async Task RunTranscodeAsync(Guid jobId)
    {
        using var scope = _scopeFactory.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();
        var job = await db.TranscodeJobs.FindAsync(jobId);
        if (job is null) return;

        var inputFile = Path.Combine(Path.GetTempPath(), $"{job.Id}-input");
        var outputFile = Path.Combine(Path.GetTempPath(), $"{job.Id}.mp4");
        try
        {
            job.Status = TranscodeStatus.Processing;
            job.UpdatedAt = DateTime.UtcNow;
            await db.SaveChangesAsync();

            await _storage.DownloadAsync(job.InputPath, inputFile);

            // Re-encode to H.264/AAC MP4 via ffmpeg as a child process.
            var psi = new ProcessStartInfo("ffmpeg")
            {
                RedirectStandardError = true,
                UseShellExecute = false
            };
            foreach (var arg in new[] { "-y", "-i", inputFile, "-c:v", "libx264", "-c:a", "aac", "-movflags", "+faststart", outputFile })
            {
                psi.ArgumentList.Add(arg);
            }
            using var proc = Process.Start(psi)!;
            var stderr = await proc.StandardError.ReadToEndAsync();
            await proc.WaitForExitAsync();
            if (proc.ExitCode != 0)
            {
                throw new Exception($"FFmpeg exited with code {proc.ExitCode}: {Truncate(stderr, 1024)}");
            }

            var objectKey = $"transcoded/{job.Id}.mp4";
            await _storage.UploadAsync(objectKey, outputFile);

            job.Status = TranscodeStatus.Completed;
            job.OutputPath = objectKey;
            job.UpdatedAt = DateTime.UtcNow;
            await db.SaveChangesAsync();
        }
        catch (Exception ex)
        {
            job.Status = TranscodeStatus.Failed;
            job.ErrorMessage = ex.Message;
            job.UpdatedAt = DateTime.UtcNow;
            await db.SaveChangesAsync();
            _logger.LogError(ex, "Transcode failed for job {JobId}", job.Id);
        }
        finally
        {
            if (File.Exists(inputFile)) File.Delete(inputFile);
            if (File.Exists(outputFile)) File.Delete(outputFile);
        }
    }

    private static string Truncate(string s, int n) => s.Length > n ? s[..n] : s;
}

File: 0-video-ingestion-and-transcoding/0-typescript/src/transcode/transcode.service.ts
import { Injectable, Logger } from "@nestjs/common"
import { InjectRepository } from "@nestjs/typeorm"
import { Repository } from "typeorm"
import { spawn } from "child_process"
import * as path from "path"
import * as fs from "fs"
import * as os from "os"
import { TranscodeJobEntity, TranscodeStatus } from "./entities/transcode-job.entity"
import { MinioService } from "./minio.service"

@Injectable()
export class TranscodeService {
    private readonly logger = new Logger(TranscodeService.name)

    constructor(
        @InjectRepository(TranscodeJobEntity)
        private readonly jobRepo: Repository<TranscodeJobEntity>,
        private readonly minioService: MinioService,
    ) {}

    /**
     * Create a PENDING job row and fire FFmpeg in the background.
     * Returns immediately — do NOT await the transcode.
     */
    async createJob(inputPath: string): Promise<TranscodeJobEntity> {
        // Idempotency: return an existing non-FAILED job for the same inputPath.
        const existing = await this.jobRepo
            .createQueryBuilder("job")
            .where("job.inputPath = :inputPath", { inputPath })
            .andWhere("job.status != :failed", { failed: TranscodeStatus.FAILED })
            .getOne()
        if (existing) return existing

        // Insert job as PENDING synchronously — caller gets a jobId immediately.
        const job = this.jobRepo.create({ inputPath, status: TranscodeStatus.PENDING })
        await this.jobRepo.save(job)

        // Fire-and-forget: run FFmpeg without blocking the HTTP response.
        this.runTranscode(job).catch((err) => {
            this.logger.error(`Unexpected error outside runTranscode for job ${job.id}`, err)
        })

        return job
    }

    async findOne(id: string): Promise<TranscodeJobEntity | null> {
        return this.jobRepo.findOne({ where: { id } })
    }

    async presignedFor(job: TranscodeJobEntity): Promise<string | null> {
        if (job.status === TranscodeStatus.COMPLETED && job.outputPath) {
            return this.minioService.presignedGetUrl(job.outputPath)
        }
        return null
    }

    /** Execute the full transcode pipeline for a single job. All errors are caught and persisted. */
    private async runTranscode(job: TranscodeJobEntity): Promise<void> {
        await this.jobRepo.update(job.id, { status: TranscodeStatus.PROCESSING })

        const tmpInput = path.join(os.tmpdir(), `${job.id}-input`)
        const tmpOutput = path.join(os.tmpdir(), `${job.id}-output.mp4`)

        try {
            await this.minioService.downloadToFile(job.inputPath, tmpInput)
            await this.spawnFfmpeg(tmpInput, tmpOutput)

            const outputPath = job.inputPath.replace(/\.[^.]+$/, "-transcoded.mp4")
            await this.minioService.uploadFromFile(tmpOutput, outputPath)

            await this.jobRepo.update(job.id, {
                status: TranscodeStatus.COMPLETED,
                outputPath,
            })
            this.logger.log(`Job ${job.id} COMPLETED -> ${outputPath}`)
        } catch (err) {
            const errorMessage = err instanceof Error ? err.message : String(err)
            await this.jobRepo.update(job.id, { status: TranscodeStatus.FAILED, errorMessage })
            this.logger.error(`Job ${job.id} FAILED: ${errorMessage}`)
        } finally {
            for (const f of [tmpInput, tmpOutput]) {
                if (fs.existsSync(f)) fs.unlinkSync(f)
            }
        }
    }

    /** Wrap FFmpeg spawn in a Promise: resolves on exit 0, rejects on non-zero exit or spawn error. */
    private spawnFfmpeg(input: string, output: string): Promise<void> {
        return new Promise((resolve, reject) => {
            const args = ["-y", "-i", input, "-vcodec", "libx264", "-acodec", "aac", "-movflags", "+faststart", output]
            const proc = spawn("ffmpeg", args)

            proc.stderr.on("data", (chunk: Buffer) => {
                this.logger.debug(`ffmpeg: ${chunk.toString().trim()}`)
            })

            proc.on("close", (code) => {
                if (code === 0) resolve()
                else reject(new Error(`FFmpeg exited with code ${code}`))
            })

            proc.on("error", (err) => {
                reject(new Error(`Failed to spawn FFmpeg: ${err.message}`))
            })
        })
    }
}
